cql3: convert term::raw to expressions

avikivity opened the pull request:
This series converts the `term::raw` objects  the grammar produces to expressions.
For each grammar production, an expression type is either reused or created.
The term::raw methods are converted to free functions accepting expressions as
input (but not yet generating expressions as output).

There is some friction because the original code base had four different expression
domains: `term`, `term::raw`, `selectable`, and `selectable::raw`. They don't match
exactly, so in some cases we need to add additional state to distinguish between them.
There are also many run-time checks introduced (on_internal_error) since the union
of the domains is much larger than each individual domain.

The method used is to erect a bi-directional bridge between term::raw and expressions,
convert various term::raw subclasses one by one, and then remove the bridge and
term::raw.


Test: unit (dev).

dekimir commented on cql3/Cql.g:
```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
+    | { sign=""; } ('-' {sign = "-"; } )? t=(K_NAN | K_INFINITY) { $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::floating_point, sign + $t.text}; }
+    ;
+
+mapLiteral returns [cql3::expr::expression map]
```
Looking at the grammar, it looks like the only use for mapLiteral is to feed its result to convert_property_map().  Can that be right?  I feel like CQL ought to allow map literals in eg, equality checks.

avikivity replied:
there is setOrMapLiteral, which is a term.

dekimir replied:
OK.  Then I suggest returning a collection_constructor from mapLiteral.

avikivity replied:
Right. Changed locally, will push after more feedback.

dekimir commented on cql3/Cql.g:
```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
-    | { s.push_back(t); }
-          ( ',' tn=term { s.push_back(tn); } )*
-      { $value = ::make_shared<cql3::sets::literal>(std::move(s)); }
+setOrMapLiteral[cql3::expr::expression t] returns [cql3::expr::expression value]
```
This return type implies that a setOrMapLiteral might be a binary operation, conjunction, or a bool, which isn't quite true.  Why not change it to collection_constructor (the lowest common ancestor node for set and map)?

dekimir replied:
I think it reads better and is more future-proof if setOrMapLiteral cannot be a binop.

avikivity replied:
Done.

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -80,11 +81,18 @@ class column_mutation_attribute;
                                 unresolved_identifier, column_mutation_attribute, function_call, cast,
-                                field_selection>;
+                                field_selection, null, bind_variable, untyped_constant,
+                                tuple_constructor, collection_constructor, usertype_constructor>;
```
I wonder if our hierarchy is too flat.  The only common ancestor for, eg, tuple and list is the root?

avikivity replied:
I don't think we'll have cases where we want to specially treat *_constructor relative to the others. They're just expressions.

dekimir replied:
And yet all the visitors added in this PR have identical code for handling tuple_constructor and collection_constructor.  OTOH, we can always add layers in the future, so this is fine for now.

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -183,14 +191,53 @@ struct function_call {
+// A constant which does not yet have a date type. It is partially typed
+// (we know if it's floating or int) but not sized.
+struct untyped_constant {
+    enum type_class { integer, floating_point, string, boolean, duration, uuid, hex };
```
Nitpick: this would be nicer as a super_enum of cql3_type::kind.

avikivity replied:
No, cql3_type::kind describes types, and this describes type classes (e.g. int/smallint/bigint/varint are all integers).

dekimir replied:
Doesn't super_enum definition allow using just a subset?

avikivity replied:
Maybe, but `integer` here doesn't map to any cql3_type::kind member. It maps to all of the integer ones.

In any case, I'd like to disappear it an replace it by a typed constant, but this series tries not to do too many changes at once. Replacing it with a typed constant will require a widening or narrowing pass later (depending on whether we represent constants in the narrowest or widest possible form).

dekimir replied:
OK.

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -183,14 +191,53 @@ struct function_call {
+    sstring raw_text;
+};
+
+// Constructs a tuple from elements
```
Not sure I understand the blurb.  The whole point of POD structs is not to act ("constructs") but to represent.  I'd like to know what `tuple_constructor` represents.

avikivity replied:
In the CQL world, this constructs a tuple. In the C++ world it's just a POD. I'll clarify the comment.

avikivity replied:
Updated both the comment and the commit changelog.

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -379,6 +480,15 @@ extern expression replace_column_def(const expression&, const column_definition*
+
+extern assignment_testable::test_result test_assignment_all(const std::vector<expression>& exprs, database& db, const sstring& keyspace, const column_specification& receiver);
+
+extern shared_ptr<assignment_testable> as_assignment_testable(expression e);
```
If there were blurbs on the method-versions of these, let's please preserve them.  If not, this is fine.

avikivity replied:
Will check and update.

avikivity replied:
Done

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -409,7 +519,30 @@ nested_expression::nested_expression(ExpressionElement auto e)
 /// Required for fmt::join() to work on expression.
-template <> struct fmt::formatter<cql3::expr::expression>;
+template <>
+struct fmt::formatter<cql3::expr::expression> {
```
@cvybhu please note :point_up:

avikivity commented on cql3/Cql.g:
```diff
@@ -219,37 +218,39 @@ struct uninitialized {
-        if (!map || map->entries.empty()) {
+    std::map<sstring, sstring> convert_property_map(const cql3::expr::expression& map_expr) {
+        auto map = std::get<cql3::expr::collection_constructor>(map_expr);
+        if (map.elements.empty()) {
```
I agree, but I wanted to avoid all unnecessary changes and make this as close as possible to a mechanical translation (of course I did make much larger changes when choosing to merge some types, for example maps::literal and sets::literal becoming collection_constructor). I'm happy to follow up with this simplification.

avikivity commented on cql3/Cql.g:
```diff
@@ -219,37 +218,39 @@ struct uninitialized {
+            auto entry_tuple = std::get_if<cql3::expr::tuple_constructor>(&entry);
             // Because the parser tries to be smart and recover on error (to
             // allow displaying more than one error I suppose), we have null
             // entries in there. Just skip those, a proper error will be thrown in the end.
```
It's partially true. We won't get a null, but we'll get a default-constructed expression (which happens to be a bool).

dekimir replied:
I'd still update the "null entries" part.

avikivity replied:
Done

avikivity commented on cql3/Cql.g:
```diff
@@ -219,37 +218,39 @@ struct uninitialized {
             if (!right) {
-                sstring msg = "Invalid property value: " + entry.first->to_string() + " for property: " + entry.second->to_string();
-                if (dynamic_pointer_cast<cql3::abstract_marker::raw>(entry.second)) {
+                sstring msg = fmt::format("Invalid property value: {} for property: ", entry_tuple->elements[0], entry_tuple->elements[1]);
```
Yes, will fix.

avikivity replied:
Done

avikivity commented on cql3/Cql.g:
```diff
@@ -501,10 +502,10 @@ insertStatement returns [std::unique_ptr<raw::modification_statement> expr]
         bool if_not_exists = false;
         bool default_unset = false;
-        ::shared_ptr<cql3::term::raw> json_value;
+        std::optional<cql3::expr::expression> json_value;
```
No, we can also end up in the non-json variation.

dekimir replied:
The non-json validation doesn't touch this variable.  The only code reading the variable dereferences it unconditionally.

avikivity replied:
Right, and therefore I think it's correct to only initialize it if we ever use it. If I remove std::optional<>, then it only compiles because we happen to have a default constructor for expressions.

dekimir replied:
OK.

avikivity commented on cql3/Cql.g:
```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
+                    if (text.size() == 1 && text[0] == '\xFF') {
+                        text = {};
+                    }
+                    $constant = cql3::expr::untyped_constant{cql3::expr::untyped_constant::string, std::move(text)}; }
```
Yes, will fix.

avikivity replied:
Done

avikivity commented on cql3/Cql.g:
```diff
@@ -1422,74 +1423,81 @@ idxName[cql3::index_name& name]
-          ( k1=term ':' v1=term { m.push_back(std::pair<shared_ptr<cql3::term::raw>, shared_ptr<cql3::term::raw>>{k1, v1}); } ( ',' kn=term ':' vn=term { m.push_back(std::pair<shared_ptr<cql3::term::raw>, shared_ptr<cql3::term::raw>>{kn, vn}); } )* )?
-      '}' { $map = ::make_shared<cql3::maps::literal>(m); }
+          ( k1=term ':' v1=term { m.push_back(cql3::expr::tuple_constructor{{std::move(k1), std::move(v1)}}); }
+            ( ',' kn=term ':' vn=term { m.push_back(cql3::expr::tuple_constructor{{std::move(kn), std::move(vn)}}); } )*  )?
```
The motivation is to allow more generic transformations to just work, in the same way that std::map::value_type is a pair and so iterators and <algorithm> can work. Things like collecting bind variables and type inference - we'll have a pass that applies the "receiver" set's value type to all elements in the collection_constructor, so if we define the map's value_type to be  tuple(key_type, value_type) (okay the names clash), then the same type inference will work for all collections.

dekimir replied:
> std::map::value_type is a pair and so iterators and can work

pair != tuple.  You're currently using tuple, and I'm suggesting using pair.

avikivity replied:
Do we want to introduce a pair just for this? It means every general-purpose tool we add will have to consider pairs too.

dekimir replied:
Why every tool?  I'd expect only tools dealing with map literals to look at map entries.

avikivity replied:
We have plenty of recursive-descent tools that operate on everything, and we'll have more for type inference. Consider

    UPDATE tab SET my_map = { 'city': ?, 'zip': ? }

this translates to (current approach)

    SET my_map = collection_constructor{{tuple_constructor{constant, bind_variable}, tuple_constructor{constant, bind_variable}}}

We now have to find the types of the bind variables. Let's say my_map is of type map<text, text>. So type inference will have to propagate the type of my_map down, checking against the constants and assigning types to bind_variable. Similarly for collect_marker_specification (now known as fill_prepare_context), or evaluate() or anything else. It's not a disaster to introduce yet another subexpression types, but we have plenty already and I think we should avoid introducing unnecessary ones.

dekimir replied:
I wouldn't make map-entry an expression; existing visitors don't need to handle it.  You're right that processing pairs and tuples will be mostly identical code, but there are other options for reuse; we don't need to check if a tuple is a pair every time we touch it.  It saps not only performance but developer's productivity, too.

dekimir replied:
> I don't think we'll need to check in many places. Most places just don't care! The only places that care are actual map construction, even even then they can trust the prepare phase to have verified it. This is not something the user can inject.

The other thing I realized is that this keeps the AST simple and flat, which has advantages.  Let's keep it as is; we can later have a use-pairs-in-map-literals RFC to illustrate the difference and allow specific discussion.

avikivity commented:
@cvybhu I'm waiting for your review too.

dekimir reviewed (commented):
Whew, finished term.cc.  I'm on (arguably well-deserved) vacation tomorrow, so I'll review the rest on Monday. :)

nyh commented on cql3/expr/expression.hh:
```diff
@@ -208,6 +209,14 @@ struct bind_variable {
 };
 
+// A constant which does not yet have a date type. It is partially typed
+// (we know if it's floating or int) but not sized.
```
So it's really strange that you call it an "untyped" constant, yet it does have a type, just not a size. So why not call it an unsized_constant?
Or at least partially_typed_constant?

avikivity replied:
It does not have a type in our type system.

nyh commented on cql3/Cql.g:
```diff
@@ -1481,17 +1481,17 @@ usertypeLiteral returns [shared_ptr<cql3::user_types::literal> ut]
-    : '(' t1=term { l.push_back(t1); } ( ',' tn=term { l.push_back(tn); } )* ')'
+tupleLiteral returns [cql3::expr::expression tt]
+    @init{ std::vector<cql3::expr::expression> l; }
+    @after{ $tt = cql3::expr::tuple_constructor{std::move(l)}; }
```
Why is the name "tuple_constructor" better than just "tuple"?

avikivity replied:
This expression constructs a tuple from several subexpressions. Many other expressions can yield a tuple, for example bind_variable or column_value. I'll add the explanation to the changelog.

avikivity replied:
Done

nyh commented on cql3/Cql.g:
```diff
@@ -1467,7 +1467,11 @@ collectionLiteral returns [shared_ptr<cql3::term::raw> value]
           ( t1=term { l.push_back(t1); } ( ',' tn=term { l.push_back(tn); } )* )?
-      ']' { $value = ::make_shared<cql3::lists::literal>(std::move(l)); }
+      ']' { $value = cql3::expr::as_term_raw(
+                        cql3::expr::collection_constructor{
```
Again, I'm curious why we don't want to just call it cql3::expr::collection() instead of collection_constructor?

avikivity replied:
It's not the only way to get a collection (selecting a collection column is another). Also, it explains that we're constructing a collection from more primitive elements.

avikivity replied:
Updated the changelog.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+static
+shared_ptr<term>
+usertype_constructor_prepare_term(const usertype_constructor& u, database& db, const sstring& keyspace, const column_specification_or_tuple& receiver_) {
+    auto& receiver = std::get<lw_shared_ptr<column_specification>>(receiver_);
```
See a509c0f634fafacd48407628f7d489cd5be6089e

dekimir replied:
I did, but I still don't understand.  a509c0f unifies prepare() methods from two old base classes (ostensibly "because we'll only have one expression type to represent both", which I don't think is even true) whereas this is a free function, unburdened by either inheritance or overloads.

avikivity replied:
> I did, but I still don't understand. [a509c0f](https://github.com/scylladb/scylla/commit/a509c0f634fafacd48407628f7d489cd5be6089e) unifies prepare() methods from two old base classes (ostensibly "because we'll only have one expression type to represent both", which I don't think is even true)

Why is it not true? We have expressions that represent scalars and tuples, in a single type.

column_specification is not exactly an expression - it's a named variable, which can be either real or manufactured (the formal argument of a function). We have the real variant of this as expressions column_value and column_value_tuple. We don't yet have a way to represent a manufactured variable, but bind_variable is pretty close.

(I'd like to eliminate column_value_tuple in favor of tuple_constuctor{{column_value, column_value...}} but that's separate)

> whereas this is a free function, unburdened by either inheritance or overloads.

a509c0f weakened the type of the receiver/receivers parameter, but kept the function a member function. This commit just changes a member function to a free function.

We can't split it to two functions with separate signatures, because the caller has the unified signature.

Ah, I think I see what you mean. multi_column_restriction.hh now creates a tuple-style receiver and directly calls prepare_term (probably the only one that does this). So we could have a parallel prepare_term implementation that accepts a tuple-style receiver, instead of folding column_specification/tuple into one type and unfolding it later. This would result in some amounts of code duplication, but probably tolerable.

Ultimately, I think the way to remove column_specification_or_tuple is to make column_specification contain an expression, and pass just a column_specification (not _or_tuple). So the _or_tupleness is determined by the receiver having CQL type tuple, rather than having an overload that takes tuple, or (as is the case now) a variant that contains either a scalar or a tuple.

avikivity replied:
To put it into concrete terms: multi_column_raw is checking that a bunch of terms are compatible with a bunch of columns:


    (x, y, z) = (2, 'rsf', 54)

the column_specification_or_tuple is the (x, y, z). expr::expression can already express (x, y, z), but it can't express typed formal variables yet. Perhaps the way to eliminate column_specification is to introduce a formal_variable expression (a type + name).

avikivity replied:
I looked at the code, and the amount of duplication will be minimal. So the question is whether we believe in the unification for column_specification[_or_tuple] into expressions or not.

dekimir replied:
> Ah, I think I see what you mean. multi_column_restriction.hh now creates a tuple-style receiver and directly calls prepare_term (probably the only one that does this). So we could have a parallel prepare_term implementation that accepts a tuple-style receiver, instead of folding column_specification/tuple into one type and unfolding it later.

Yes, that's what I meant.

> So the question is whether we believe in the unification for column_specification[_or_tuple] into expressions or not.

It seems clear that in this PR, the answer is still "no."

avikivity replied:
Obviously the series doesn't do that. The question is about the direction. If it's a good direction, then splitting prepare_term() into prepare_term() and prepare_term_multi_column() will be wasted work since we'll have to reunify them (replacing the column_specification_or_tuple with a single expression). If it's not a good direction, then splitting prepare_term() removes nasty variants and improves the code.

dekimir replied:
Seems to me like column_specification is little more than a type.  It's just used to test assignability.  I can't even find any column-name usage, outside error messages.

If types aren't expressions, neither are column_specifications.

avikivity replied:
It's like a function formal parameter. A type accompanied by a name which is only good for error reporting.

A formal parameter isn't an expression, but it becomes part of a fake expression when the compiler assigns it an actual argument.

Right now column_specification_or_tuple can denote
 - a function formal parameter
 - a table column
 - a tuple of table columns
 - a parameter to a CQL clause (LIMIT ...)
 - tuple elements
 - user defined type element

But, it's probably faster for me to split prepare_term() than to continue arguing, and reunify it when the time comes.

avikivity replied:
Actually, I think splitting prepare_term() can reduce duplication! Let me see if it works out.

avikivity replied:
Done, the two new patches in the end.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+
+lw_shared_ptr<column_specification>
+user_types::field_spec_of(const column_specification& column, size_t field) {
+    return cql3::expr::usertype_field_spec_of(column, field);
```
Yes, it isn't pretty. Most of the raw domain function disappear, but a few remain. I expect more to disappear.

As an example, this particular method will disappear when we convert operation::set_field to a field_reference{column_value, identifier} expression.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+/*
+ * Copyright (C) 2020-present ScyllaDB
```
Fixed.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+
+namespace cql3::expr {
+
+static
```
I dislike anonymous namespaces since they bloat call stacks. I'd hate seeing `cql3::expr::{anonymous namespace}::whatever`.

dekimir replied:
Coder's choice, keep it.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+    }
+
+    auto&& value_spec = set_value_spec_of(receiver);
+    // FIXME: make assignment_testable::test_all() accept ranges
```
Actually, we can remove the FIXME, since by the end of the series, we don't need test_all to accept generic ranges.

avikivity replied:
Done (in cql3: expr: don't convert subexpressions to term::raw during the prepare phase)

avikivity commented:
v2: update with all the fixes I agreed with:
- make mapLiteral grammar production return collection_constructor
- adjustd comment in convert_property_map() not to refer to null entries any more (Cql.g)
- fixed format string in convert_property_map() (Cql.g)
- fixed minor formatting in C++ code in constant production (Cql.g)
- make setOrMapLiteral production return a collection_constructor instead of a generic expression (Cql.g)
- improved comment and changelog on tuple_constructor
- carry over comments on test_assignment_all() and test_assignment()
- adjusted copyright year on term-expr.cc (which is deleted by the end of the series!)
- remove FIXME in set_test_assignment() which became outdated
- updated changelog for collection_constructor to explain the name choice

cvybhu commented:
It looks like this branch doesn't compile. I get an error:
```bash
[jancio@localhost]~/code/scylla$ ninja build/dev/cql3/expr/expression.o
[1/2] NINJA build/dev/seastar/libseastar.a
ninja: Entering directory `build/dev/seastar'
ninja: no work to do.
[2/2] CXX build/dev/cql3/expr/expression.o
FAILED: build/dev/cql3/expr/expression.o 
clang++ -MD -MT build/dev/cql3/expr/expression.o -MF build/dev/cql3/expr/expression.o.d -I/home/jancio/code/scylla/seastar/include -I/home/jancio/code/scylla/build/dev/seastar/gen/include -std=gnu++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=6 -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -iquote. -iquote build/dev/gen --std=gnu++20  -ffile-prefix-map=/home/jancio/code/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-unknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/dev/cql3/expr/expression.o cql3/expr/expression.cc
In file included from cql3/expr/expression.cc:39:
In file included from ./cql3/selection/selection.hh:49:
./cql3/selection/raw_selector.hh:46:10: fatal error: 'cql3/selection/selectable-expr.hh' file not found
#include "cql3/selection/selectable-expr.hh"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
distcc[133619] ERROR: compile cql3/expr/expression.cc on localhost failed
ninja: build stopped: subcommand failed.
```

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -183,14 +191,53 @@ struct function_call {
+    struct tuple {};
+    struct tuple_in {};
+    // FIXME: infer shape from expression rather than from grammar
+    std::variant<scalar, scalar_in, tuple, tuple_in> shape;
```
Variant of empty structs seems strange, why not an `enum`?

avikivity replied:
I thought we might later have need for a real variant. Instead, it will probably disappear (and shape will be inferred using types instead of the grammar). I'll change it to enum.

avikivity replied:
Fixed

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -379,6 +480,27 @@ extern expression replace_column_def(const expression&, const column_definition*
 extern expression replace_token(const expression&, const column_definition*);
 
+
+extern ::shared_ptr<term> prepare_term(const expression& expr, database& db, const sstring& keyspace, const column_specification_or_tuple& receiver);
```
`extern` is not needed in `expression.hh`

avikivity replied:
Last time I didn't add it @dekimir complained, so I added it.

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -183,14 +191,53 @@ struct function_call {
+
+// Constructs an object of a user-defined type
+struct usertype_constructor {
+    using elements_map_type = std::unordered_map<column_identifier, nested_expression>;
```
It looks like `std::map` works with incomplete types and doesn't require `nested_expression`. The code would be simpler, but maybe slower (?)

avikivity replied:
I think going with unordered_map is better since that's what user_types::literal used. Later we can consider switching to std::map. I don't think there's much difference, since user types don't have enough members to make the container complexity important.

OTOH I don't think we have ordering for column_identifier, so std::map won't work out of the box.

cvybhu commented on cql3/selection/raw_selector.hh:
```diff
@@ -43,6 +43,7 @@
 #pragma once
 
 #include "cql3/selection/selectable.hh"
+#include "cql3/selection/selectable-expr.hh"
```
This file doesn't exist

cvybhu reviewed (commented):
Went through some files, will continue tomorrow

avikivity commented:
> It looks like this branch doesn't compile. I get an error:
> 
> ```shell
> [jancio@localhost]~/code/scylla$ ninja build/dev/cql3/expr/expression.o
> [1/2] NINJA build/dev/seastar/libseastar.a
> ninja: Entering directory `build/dev/seastar'
> ninja: no work to do.
> [2/2] CXX build/dev/cql3/expr/expression.o
> FAILED: build/dev/cql3/expr/expression.o 
> clang++ -MD -MT build/dev/cql3/expr/expression.o -MF build/dev/cql3/expr/expression.o.d -I/home/jancio/code/scylla/seastar/include -I/home/jancio/code/scylla/build/dev/seastar/gen/include -std=gnu++20 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=6 -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_TYPE_ERASE_MORE -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -O2 -iquote. -iquote build/dev/gen --std=gnu++20  -ffile-prefix-map=/home/jancio/code/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -Iabseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-inconsistent-missing-override -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-unknown-attributes -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/dev/cql3/expr/expression.o cql3/expr/expression.cc
> In file included from cql3/expr/expression.cc:39:
> In file included from ./cql3/selection/selection.hh:49:
> ./cql3/selection/raw_selector.hh:46:10: fatal error: 'cql3/selection/selectable-expr.hh' file not found
> #include "cql3/selection/selectable-expr.hh"
>          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> distcc[133619] ERROR: compile cql3/expr/expression.cc on localhost failed
> ninja: build stopped: subcommand failed.
> ```

Oops, git add fail. Pushed a fix.

avikivity commented:
v3: added missing file selectable-expr.hh.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -848,6 +902,24 @@ value_set possible_lhs_values(const column_definition* cdef, const expression& e
                 on_internal_error(expr_logger, "possible_lhs_values: a field selection cannot serve as a restriction by itself");
             },
+            [] (const null&) -> value_set {
+                on_internal_error(expr_logger, "possible_lhs_values: a raw term cannot serve as a restriction by itself");
```
yes

avikivity replied:
Fixed

avikivity commented on cql3/Cql.g:
```diff
@@ -1506,17 +1514,17 @@ allowedFunctionName returns [sstring s]
+term returns [cql3::expr::expression term1]
+    : v=value                          { $term1 = std::move(v); }
+    | f=functionName args=functionArgs { $term1 = cql3::expr::function_call{std::move(f), std::move(args)}; }
+    | '(' c=comparatorType ')' t=term  { $term1 = cql3::expr::cast{cql3::expr::expression(t), c}; }
```
Fixed (and also removed the needless call to the expression copy constructor here in the next-to-last patch)

avikivity commented:
v4:
- add std::move() for type cast grammar->C++ glue
- convert bind_variable::shape from a variant of four empty structs to an enum
- corrected one on_internal_error() message on unexpected null

cvybhu commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+        assert(0);
+    }
+
+    if (receiver->type->is_user_type()) {
```
looks like it shoud be `receiver->type->without_reversed().is_user_type()`

dekimir replied:
Copied verbatim from old code.  @cvybhu can you please check if master has a bug here?

cvybhu replied:
The same code is on master. It looks like a bug, but maybe it works somehow, will have to check.

cvybhu reviewed (changes requested):
Looks good,
Now that I have much better understanding of this change, I will try to make my PR more in spirit of simplifying everything to just `expression`.

avikivity commented on cql3/expr/term_expr.cc:
```diff
@@ -0,0 +1,1004 @@
+static
+void
+usertype_constructor_validate_assignable_to(const usertype_constructor& u, database& db, const sstring& keyspace, const column_specification& receiver) {
+    if (!receiver.type->is_user_type()) {
```
Probably true, but this exists in the original code too, no?

I don't want to mix a bug fix with this series, it will have to be separate.

cvybhu replied:
It does, I will open an issue.

dekimir replied:
When you open it, please reference it here.

cvybhu replied:
Issue #9215

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -965,11 +1055,84 @@ std::ostream& operator<<(std::ostream& os, const expression& expr) {
+                case collection_constructor::style_type::set: {
+                    fmt::print(os, "{{");
+                    bool first = true;
+                    for (auto& e : cc.elements) {
```
We could.

avikivity replied:
And now we do.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -965,11 +1055,84 @@ std::ostream& operator<<(std::ostream& os, const expression& expr) {
+            },
+            [&] (const bind_variable&) {
+                // FIXME: store and present bind variable name
+                fmt::print(os, "?");
```
Any changes to the string form trigger failures in unit tests, so these need to done carefully and outside this large change.

As an example, printing null as NULL causes tests to fail.

:{bind_index} is wrong, since it means we can't round-trip CQL.

dekimir replied:
OK.

avikivity commented:
> I wouldn't make map-entry an expression; existing visitors don't need to handle it. You're right that processing pairs and tuples will be mostly identical code, but there are other options for reuse; we don't need to check if a tuple is a pair every time we touch it. It saps not only performance but developer's productivity, too.

I don't think we'll need to check in many places. Most places just don't care! The only places that care are actual map construction, even even then they can trust the prepare phase to have verified it. This is not something the user can inject.

avikivity commented:
v5:
 - use fmt::join() to format set elements of collection_constructor

avikivity commented:
@psarna please review. We weren't able to agree on whether collection_constructor for maps should use tuples for elements or a new expression type, so you get to make that call.

dekimir commented:
> We weren't able to agree on whether collection_constructor for maps should use tuples for elements or a new expression type

I thought we did; I [yielded](https://github.com/scylladb/scylla/pull/9170/files#r691165595).

dekimir commented:
I also don't see the response to these comments:
* [is column_specification_or_tuple still necessary in the end](https://github.com/scylladb/scylla/pull/9170/files#r689993220)?
* [null -> bind_variable](https://github.com/scylladb/scylla/pull/9170/files#r690211132)

avikivity commented on cql3/selection/selectable.cc:
```diff
@@ -195,14 +195,37 @@ prepare_selectable(const schema& s, const expr::expression& raw_selectable) {
+            on_internal_error(slogger, "null found its way to selector context");
+        },
+        [&] (const expr::bind_variable&) -> shared_ptr<selectable> {
+            on_internal_error(slogger, "null found its way to selector context");
```
Fixed.

avikivity commented:
Sorry, missed them. Responded to one and fixed the other.

avikivity commented:
v6:
 - fix copy-paste reporting a bind_variable in internal error case as null

avikivity commented:
v7:
  - remove column_specification_or_tuple as the means to prepare something that is compared to a tuple, instead add a new free function prepare_term_multi_column() (two new patches at the end)

dekimir commented on cql3/expr/term_expr.cc:
```diff
@@ -875,6 +852,27 @@ prepare_term(const expression& expr, database& db, const sstring& keyspace, cons
+        [&] (const tuple_constructor& tc) -> ::shared_ptr<term> {
+            return tuple_constructor_prepare_tuple(tc, db, keyspace, receivers);
+        },
+        [] (const auto& default_cast) -> ::shared_ptr<term> {
```
What a great demo of `default_case` usefulness, despite the typo. ;)

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -965,11 +1055,75 @@ std::ostream& operator<<(std::ostream& os, const expression& expr) {
+            },
+            [&] (const usertype_constructor& uc) {
+                fmt::print(os, "{{");
+                bool first = true;
```
join()?

avikivity replied:
This can't easily be joined, because we print the field name and value independently.
